src/max_weight_matching.rs

```diff
@@ -0,0 +1,1306 @@
+    // If b is a non-trivial (sub-)blossom,
+    // blossom_children[b] is an ordered list of its sub-blossoms, starting with
+    // the base and going round the blossom.
+    let mut blossom_children: Vec<Vec<usize>> =
+        (0..2 * num_nodes).map(|_| Vec::new()).collect();
+    // If b is a (sub-)blossom,
+    // blossombase[b] is its base VERTEX (i.e. recursive sub-blossom).
+    let mut blossom_base: Vec<Option<usize>> =
+        (0..num_nodes).map(Some).collect();
+    blossom_base.append(&mut vec![None; num_nodes]);
+    // If b is a non-trivial (sub-)blossom,
+    // blossom_endpoints[b] is a list of endpoints on its connecting edges,
+    // such that blossom_endpoints[b][i] is the local endpoint of
+    // blossom_children[b][i] on the edge that connects it to
+    // blossom_children[b][wrap(i+1)].
+    let mut blossom_endpoints: Vec<Vec<usize>> =
+        (0..2 * num_nodes).map(|_| Vec::new()).collect();
+    // If v is a free vertex (or an unreached vertex inside a T-blossom),
+    // best_edge[v] is the edge to an S-vertex with least slack,
+    // or None if there is no such edge. If b is a (possibly trivial)
+    // top-level S-blossom,
+    // best_edge[b] is the least-slack edge to a different S-blossom,
+    // or None if there is no such edge.
+    let mut best_edge: Vec<Option<usize>>;
+    // If b is a non-trivial top-level S-blossom,
+    // blossom_best_edges[b] is a list of least-slack edges to neighboring
+    // S-blossoms, or None if no such list has been computed yet.
+    // This is used for efficient computation of delta3.
+    let mut blossom_best_edges: Vec<Vec<usize>> =
+        (0..2 * num_nodes).map(|_| Vec::new()).collect();
+    let mut unused_blossoms: Vec<usize> = (num_nodes..2 * num_nodes).collect();
+    // If v is a vertex,
+    // dual_var[v] = 2 * u(v) where u(v) is the v's variable in the dual
+    // optimization problem (multiplication by two ensures integer values
+    // throughout the algorithm if all edge weights are integers).
+    // If b is a non-trivial blossom,
+    // dual_var[b] = z(b) where z(b) is b's variable in the dual optimization
+    // problem.
+    let mut dual_var: Vec<i128> = vec![max_weight; num_nodes];
+    dual_var.append(&mut vec![0; num_nodes]);
+    // If allowed_edge[k] is true, edge k has zero slack in the optimization
+    // problem; if allowed_edge[k] is false, the edge's slack may or may not
+    // be zero.
+    let mut allowed_edge: Vec<bool>;
+    // Queue of newly discovered S-vertices
+    let mut queue: Vec<usize> = Vec::with_capacity(num_nodes);
+
+    // Main loop: continue until no further improvement is possible
+    for _ in 0..num_nodes {
+        // Each iteration of this loop is a "stage".
+        // A stage finds an augmenting path and uses that to improve
+        // the matching.
+
+        // Removal labels from top-level blossoms/vertices
+        labels = vec![Some(0); 2 * num_nodes];
```

Comment:
Yeah, I think that's good to do probably, I'll add this.

I think it might not be strictly necessary here because we end up reset `label_ends` every time we update `labels`. But to be more thorough we should reset the endpoints when we reset labels.

Reply:
Done in: https://github.com/Qiskit/retworkx/pull/229/commits/81a59a491d637f3600456b23f84a3662f0eccb2f